### ISSUE TITLE
On list view total is calculated on entire table, not applied scope or where

### ISSUE THREAD

**Issue by tkerns** [Laravel-Backpack/CRUD#2498]
# Bug report

### What I did
On the list view, if you apply a scope, or a where in the list setup like this
`$this->crud->addClause('active'); // apply a local scope`
and before I have filtered the record set in the UI,

### What I expected to happen

I would expect it to show something like this at the top.

**Showing 1 to 10 of 730 entries**

and then when I filter using the search box, or a filter, for it to tell me the results have been filtered like this

**Showing 1 to 10 of 250 entries (filtered from 730 total entries)**

### What happened

Instead, it show me this:

**Showing 1 to 10 of 730 entries (filtered from 1,021 total entries)**

even before filtering, just because I applied the scope in the setupListOperation

When I run ```php artisan backpack:version``` the output is:
### PHP VERSION:
PHP 7.3.14 (cli) (built: Jan 21 2020 13:15:55) ( ZTS MSVC15 (Visual C++ 2017) x64 )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.3.14, Copyright (c) 1998-2018 Zend Technologies

### LARAVEL VERSION:
laravel/framework                     v6.17.1            The Laravel Framework.

### BACKPACK VERSION:
backpack/crud                         4.0.41             Quickly build an admin interfaces using Laravel 6, CoreUI...
backpack/generators                   2.0.6              Generate files for laravel projects

**Comment by tabacitu** [comment]
Thank you for moving the issue from Reddit to Github @tkerns !

## Problem

This is because inside ```ListOperation::search()```, we're doing:
```php
        // ..
        $totalRows = $this->crud->model->count();
        $filteredRows = $this->crud->count();
        // ...
```

If you were using a [Global Scope](https://laravel.com/docs/6.x/eloquent#global-scopes), the numbers would have been ok. But since you're using a local scope, ```$totalRows``` won't show what you're looking for, only the active users number.

## Solution

We could provide a way around this issue, by adding a new method inside CrudTrait, something like 
```php
public function getCountForCrud() {
    return $this->count();
}
```

and calling that method inside ```ListOperation::search()``` instead:

```diff
        // ..
-       $totalRows = $this->crud->model->count();
+       $totalRows = $this->crud->model->getCountForCrud();
        $filteredRows = $this->crud->count();
        // ...
```

People would then be able to easily overwrite this method in their model:
```php
public function getCountForCrud() {
    return $this->active()->count();
}
```

That would work. But... I don't know how I feel about the change... Apart from the bad naming, it's kind of weird to define something inside the Model that's only specific to the CRUD. It makes the Model less general...

Anybody else - thoughts? A better solution?

## Temporary Solution

Let's wait for others to pitch in with their opinion @tkerns  - maybe someone else has a better long-term solution. Until then, you can overwrite the ```search()``` method in your controller to do this manually:
```diff

    /**
     * The search function that is called by the data table.
     *
     * @return array JSON Array of cells in HTML form.
     */
    public function search()
    {
        $this->crud->hasAccessOrFail('list');

-       $totalRows = $this->crud->model->count();
+       $totalRows = $this->crud->model->active()->count();
        $filteredRows = $this->crud->count();
        $startIndex = $this->request->input('start') ?: 0;
        // if a search term was present
        if ($this->request->input('search') && $this->request->input('search')['value']) {
            // filter the results accordingly
            $this->crud->applySearchTerm($this->request->input('search')['value']);
            // recalculate the number of filtered rows
            $filteredRows = $this->crud->count();
        }
        // start the results according to the datatables pagination
        if ($this->request->input('start')) {
            $this->crud->skip((int) $this->request->input('start'));
        }
        // limit the number of results according to the datatables pagination
        if ($this->request->input('length')) {
            $this->crud->take((int) $this->request->input('length'));
        }
        // overwrite any order set in the setup() method with the datatables order
        if ($this->request->input('order')) {
            $column_number = $this->request->input('order')[0]['column'];
            $column_direction = $this->request->input('order')[0]['dir'];
            $column = $this->crud->findColumnById($column_number);
            if ($column['tableColumn']) {
                // clear any past orderBy rules
                $this->crud->query->getQuery()->orders = null;
                // apply the current orderBy rules
                $this->crud->query->orderBy($column['name'], $column_direction);
            }

            // check for custom order logic in the column definition
            if (isset($column['orderLogic'])) {
                $this->crud->customOrderBy($column, $column_direction);
            }
        }

        // show newest items first, by default (if no order has been set for the primary column)
        // if there was no order set, this will be the only one
        // if there was an order set, this will be the last one (after all others were applied)
        $orderBy = $this->crud->query->getQuery()->orders;
        $hasOrderByPrimaryKey = false;
        collect($orderBy)->each(function ($item, $key) use ($hasOrderByPrimaryKey) {
            if (! isset($item['column'])) {
                return false;
            }

            if ($item['column'] == $this->crud->model->getKeyName()) {
                $hasOrderByPrimaryKey = true;

                return false;
            }
        });
        if (! $hasOrderByPrimaryKey) {
            $this->crud->query->orderByDesc($this->crud->model->getKeyName());
        }

        $entries = $this->crud->getEntries();

        return $this->crud->getEntriesAsJsonForDatatables($entries, $totalRows, $filteredRows, $startIndex);
    }
```
Or... you know... make it a Global Scope if that makes sense in your app.

**Comment by tkerns** [comment]
@tabacitu, unfortunately, a global scope will not work in this situation, but that workaround will do just fine, thank you!  I agree that putting something in the model that relates only to crud doesn't feel quite right.  Although I have done that for other things like calling model methods for crud fields.  However, yes, lets wait and see if a nice elegant solution comes up.  I will think about it some as well.

**Comment by tabacitu** [comment]
Just though of another temporary solution for you @tkerns . You could:
- create a new model, say ```ActiveProduct```, which extends the Product model AND has the global scope;
- make sure the new model uses this trait ```Backpack\CRUD\app\Models\Traits\InheritsRelationsFromParentModel;``` to inherit Eloquent relationships;
- use ```ActiveProduct``` instead of ```Product``` in your ProductCrudController;

That way you don't have to overwrite the operation at all, because you're using a global scope. I think this one's better.

**Comment by tkerns** [comment]
@tabacitu Yes, that would work in that situation.  However in 2 other situations where I have encountered it, it wouldn't work.  One is where the crud is a nested crud, and get its base list based on url parameters.  Another is where the crud list is based on the authenticated user (like posts belonging to a user).  I guess it just seems like it the list should get its base count of total unfiltered models based on any where clauses or scopes in the setup method.   I think for now, until  a solution comes along that does that, I will probably just overwrite the search method.
Thanks for the idea though, as I do have one situation where that would work fine.

**Comment by zachweix** [comment]
I've actually had this issue on projects, and the way that I've handled it is defining on the controller:
```
protected function recordsTotalOverride() {
    $count = $this->crud->model->where(`insert_whatever_you_want_here`)->count();
    return $count;
}
```
On the parent Controller I added the same function where `$count` is calculated the same way it is now in the current search function (`$totalRows = $this->crud->model->count();`)

And then I rewrote the `search()` function to look like this:

```
/**
 * The search function that is called by the data table.
 *
 * @return array JSON Array of cells in HTML form.
 */
public function search()
{
    $this->crud->hasAccessOrFail('list');
    $this->crud->setOperation('list');

    $totalRows = $this->crud->model->count();
    $filteredRows = $this->recordsTotalOverride();
    $startIndex = $this->request->input('start') ?: 0;
    ...
```
